Chatto/sources/ChatController/ChatMessages/New/ChatCollectionViewLayoutModelFactory.swift

```diff
@@ -0,0 +1,84 @@
+/*
+ The MIT License (MIT)
+
+ Copyright (c) 2015-present Badoo Trading Limited.
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+import UIKit
+
+public protocol ChatCollectionViewLayoutModelFactoryProtocol: AnyObject {
+    func createLayoutModel(items: ChatItemCompanionCollection,
+             collectionViewWidth: CGFloat) -> ChatCollectionViewLayoutModel
+}
+
+final class ChatCollectionViewLayoutModelFactory: ChatCollectionViewLayoutModelFactoryProtocol {
+
+    // MARK: - Instantiation
+
+    init() {
+    }
+
+    // MARK: - ChatCollectionViewLayoutUpdater
+
+    public func createLayoutModel(items: ChatItemCompanionCollection, collectionViewWidth: CGFloat) -> ChatCollectionViewLayoutModel {
+        // swiftlint:disable:next nesting
+        typealias IntermediateItemLayoutData = (height: CGFloat?, bottomMargin: CGFloat)
+        typealias ItemLayoutData = (height: CGFloat, bottomMargin: CGFloat)
+        // swiftlint:disable:previous nesting
+
+        func createLayoutModel(intermediateLayoutData: [IntermediateItemLayoutData]) -> ChatCollectionViewLayoutModel {
+            let layoutData = intermediateLayoutData.map { (intermediateLayoutData: IntermediateItemLayoutData) -> ItemLayoutData in
+                return (height: intermediateLayoutData.height!, bottomMargin: intermediateLayoutData.bottomMargin)
+            }
+            return ChatCollectionViewLayoutModel.createModel(collectionViewWidth, itemsLayoutData: layoutData)
+        }
+
+        let isInBackground = !Thread.isMainThread
+        var intermediateLayoutData = [IntermediateItemLayoutData]()
+        var itemsForMainThread = [(index: Int, itemCompanion: ChatItemCompanion)]()
+
+        for (index, itemCompanion) in items.enumerated() {
```

Comment:
Just an idea how we could potentially avoid the force unwrap :) 
What about the `IntermediateItemLayoutData` is equal to `(heightProvider: () -> CGFloat, bottomMargin: CGFloat)` then:
* In the `isInBackground || itemCompanion.presenter.canCalculateHeightInBackground` we would calculate it immediately and return the value
* In the `else` we would return the closure to calculate

In the end we would just run a `map` to perform make the `ItemLayoutData` in the main thread if needed